### ISSUE TITLE
refactor controllers to do partial reconcilation on watch events 

### ISSUE THREAD

**Issue by murali-reddy** [cloudnativelabs/kube-router#997]
**TLDR**: kube-router leverages Kubernetes controller pattern and watches API objects and performs reconcilation (node level changes to reflect desired state). In the current design of kube-router it does full reconcilation (force-refresh the whole state), rather than just doing incremental delta configuration update corresponding to watch API object event. This approach simplifies the design but can result in scalability issues. This proposal is to refactor current controllers to restrict the processing done on watch API object event to just scope of the change needed rather full reconcilation of desired state.

**Problem**:

kube-router is primarily consists of below three contollers

- `network policy controller` which provides an implementation of network policy and watches following Kubernetes resources
    - Pod
    - Namespace
    - NetworkPolicy
- `network service controller` which provides an implementation of service proxy and watches following Kubernetes resources
    - Service
    - Endpoint
- `network routing controller` which provides pod-to-pod connectivity also advertising the service VIP's to provide network load balancer and watches following Kubernetes resources
    - Node
    - Service
    - Endpoint

each of these controllers registers a callback to Add/Delete/Update events. At present kube-router follows simple model in which each controller performs full reconcilation irrespective of the resource object for which event was received. To safeguard against cases where there is flood of events for e.g. on node restart, deleting a deployment with large number of pods etc, kube-router attempts to coalesce the events and does the minimum syncs. This approach works faily well for most deployments. So far users feedback over last few years indicated below two main problems:

- full sync time can be unacceptle. In some of the deployments for instance network policy controller full sync is found to be in the order of couple of minutes
- if there is constant churn of resources (pods, services etc) then even with coalescing kube-router can end up continuously processing the events. Combined with longer full sync's kube-router can potentially hog the system

In full syncs if it only involves system calls its pretty fast. For e.g. `network service controller` largley invokes netlink, ipvs system calls. But inherently still full sync can not scale as number of services and endpoints increases. When there is a necessasity to call external command (resulting in process context switch) like in case of `network policy controller` which relies on `iptables` command performing full sync can be expensive. As the number of network policies and corresponding pods to which policies appy increases the number of iptables calls increases linearly.

So its important to reconsider current design to strike a balance between simplicity of design and scalability

** proposed solutions **

TODO: will be adding the details

**Comment by aauren** [comment]
It may be worth noting that `network routing controller` already does a form of partial reconciliation. I'm sure that there is still additional work that we could do there to make it more performant, but it isn't all that bad as compared to `network services controller` which is only a small step away from a full sync for IPVS services or `network policy controller` which doesn't have any notions of partial reconciliation.

**Comment by aauren** [comment]
Closing, see comment on #1010 for reason
